Sources/Datadog/Core/Persistence/Privacy/ConsentAwareDataWriter.swift

```diff
@@ -20,23 +20,61 @@ internal class ConsentAwareDataWriter: FileWriterType {
     private var activeFileWriter: FileWriterType?
 
     init(
-        initialConsent: TrackingConsent,
+        consentProvider: ConsentProvider,
         queue: DispatchQueue,
         unauthorizedFileWriter: FileWriterType,
         authorizedFileWriter: FileWriterType
     ) {
         self.queue = queue
         self.unauthorizedFileWriter = unauthorizedFileWriter
         self.authorizedFileWriter = authorizedFileWriter
+        self.activeFileWriter = resolveActiveFileWriter(
+            for: consentProvider.currentValue,
+            unauthorizedWriter: unauthorizedFileWriter,
+            authorizedWriter: authorizedFileWriter
+        )
 
-        switch initialConsent {
-        case .granted: self.activeFileWriter = authorizedFileWriter
-        case .notGranted: self.activeFileWriter = nil
-        case .pending: self.activeFileWriter = unauthorizedFileWriter
-        }
+        consentProvider.subscribe(consentSubscriber: self)
     }
 
+    // MARK: - FileWriterType
+
     func write<T>(value: T) where T: Encodable {
-        activeFileWriter?.write(value: value)
+        synchronized {
+            activeFileWriter?.write(value: value)
+        }
+    }
+
+    // MARK: - ConsentSubscriber
+
+    func consentChanged(from oldValue: TrackingConsent, to newValue: TrackingConsent) {
+        synchronized {
+            activeFileWriter = resolveActiveFileWriter(
+                for: newValue,
+                unauthorizedWriter: unauthorizedFileWriter,
+                authorizedWriter: authorizedFileWriter
+            )
+        }
+    }
+
+    // MARK: - Private
+
+    private func synchronized(block: () -> Void) {
+        objc_sync_enter(self)
+        block()
+        objc_sync_exit(self)
```

Comment:
this is totally fine 👍 i'm just wondering why we didn't use `self.queue`?

Reply:
It's because in this PR the `FileWriter` is still async. It dispatches writes asynchronously to this `queue`. Making `self.queue.async {}` in this PR will lead to doubling the async call to `queue` (`queue.async { queue.async { /* write */ } }`).

But I'm aware of this and this was changed in #330 where `queue` is used instead of `objc_sync`. In #330 the `FileWriter` is changed to be synchronous, which provides simpler threading model.